### PR TITLE
Restore OS390-Unix

### PR DESCRIPTION
**Pull request by levitte**
Removing the script used by the config target was a mistake to begin with.

### CODE REVIEW

**Review by t-j-h — APPROVED** [review]

**Comment by levitte** [comment]
I realised that the placement of the script as well as its use didn't quite fit the way OpenSSL is now built, so I added an adjusting commit.  Reconfirm?

**Review by t-j-h — APPROVED** [review]

**Comment by richsalz** [comment]
-1.  Earlier removal might have been a mistake -- emphasis on *might*, we really need @dot-asm to comment on this.  But the author of the config and script could not be reached for our CLA.

if we *need* this for OS390, and I am not sure that we do (again, ask @dot-asm) then we can do that.  But re-adding non-CLA code is a big red flag.

DO NOT DO THIS.

**Comment by levitte** [comment]
Ok...  would a rewrite suite you better, @richsalz?

**Comment by levitte** [comment]
Re-opening, because the issue you have, @richsalz, applies to master only, as the relicense effort is only applicable there.  So note that I've removed the "master" lable and only kept the "1.1.0" one.  This PR should be applied on 1.1.0.

**Comment by levitte** [comment]
(I have a completely different OS390-Unix change coming for master)

**Comment by dot-asm** [comment]
I don't quite understand why am I being called on this. The fact that I write s390x assembly doesn't mean that I know anything about OS/390. But all right... Question is why? Why is it there at all? Has *anybody* actually tested it? Is it established fact that it goes through compilation? And passes make test? Even if somebody can confirm this, wouldn't it be more appropriate to place it in separate .conf file? To denote that we have no capacity to actually maintain it. Actually nobody does, OS/390 is no more, it's superseded by z/OS many years ago. So vendor is out of picture, and I find it hard to imagine that there is enthusiast community vibrant enough to pull it off. Also note that in both cases we are talking about POSIX compatibility layer, i.e. non-native applications [whatever it means.] Why is this significant? z/OS is likely to have compatibility layer to run OS/390 *native* applications, but I doubt that it applies to compatibility layer. No matter how you turn it, this config line has no practical value.

To summarize. Do we support Windows 9x? And why not? Now replace "Windows 9x" in everything you say/think with "OS/390". If there is need for contemporary mainframe support, we would be better off doing it from scratch. And keep in mind that as long as somebody else does it, suggestion still is separate .conf file...

**Comment by richsalz** [comment]
@dot-asm you don't know why you were asked, but you provided *exactly* the right information: OS/390 is obsolete, we don't have a way to build and test it, and z/OS is the current and supported.
Close those and spend no cycles unless someone on the team has an OS390 mainframe lying around.

**Comment by levitte** [comment]
**For 1.1.0**, we've deemed it inappropriate to remove config targets.  I'm sure we have half a ton of them them are buggy, so we fix them when we can (like I did with the uClinux targets) or someone contributes a fix.  So following this pattern, OS390-Unix *should* remain in 1.1.0, and it *should* be fixed when someone gets around to it.

This is the pattern we've followed so far and that is consistent with promises we make re compatibility.  I see no reason to deviate from this pattern in this particular case.

**Comment by richsalz** [comment]
I don't remember any such decision, but that could be my memory playing trick.  I don't care much about what we do for 1.1.0 in this area however.

**Comment by dot-asm** [comment]
> For 1.1.0, we've deemed it inappropriate to remove config targets.

Life is a trade-off. It was rather inability to figure out/confirm status for *each* platform at the time as a way to *postpone* the solution. This kind of implies that there are no firm rules. Besides, you really can't do better than availability of platform. I mean if platform is unavailable, then it's really no point pretending it's supported.

**Comment by t-j-h** [comment]
We have a platform label OS390-Unix - which is historical. OS/390 support was officially ended by IBM in 2004 (yes that recently) and OS390 was the first mainframe environment to support the z-Architecture.  OS/390 is simply one of **many** names for the same logical underlying operating system (MVS) which has had many names as it was extended (by program products) since 1974. 

**_The MVS core remains fundamentally the same operating system. By design, programs written for MVS run on z/OS without modification._**

So stop focusing on the label we use in the platform definition. 

We have a large collection of platforms in our Configure/Configurations which are in exactly the same situation - where we don't regularly test them.

There is a large active community of OpenSSL users on MVS/OS-390/z-OS/System-Z/whatever label you want to call it.

I compiled up a 1.0.2 version of this environment yesterday with some minor adjustments (makefile issues) and changing from c89.sh to cc.sh as a wrapper for the -L handling (given that c89 does not support long long) and am up to make test failing in sha256 which remains to be investigated. I will compile a 'master' version shortly and have a look at it too.

**Comment by dot-asm** [comment]
The emphasized text refers to "native" application code, something we *never* supported. Target was always POSIX compatibility layer. Well, one can argue that they, POSIX layers, are similar enough to be viewed as one platform, and that they can be described by common config. [And in the process not utilize 64-bit-ness of z/OS.] But it doesn't change the fact build was disfuncitonal for many years across multiple releases. And this speaks against suggestion that there is "large active community of OpenSSL users on MVS/OS-390/z-OS". As they apparently remained silent for all those years and releases. BTW, do note that I've omitted "System-Z" from quote. It's not OS, but collective name for line of hardware. One can run OS other than MVS/OS-390/z-OS, apparently more popular one, and it's not exactly appropriate to count linux-s390x users as members of MVS/OS-390/z-OS community.

Once again. assertion is that we would be better off considering this problem as *new*. There is no evidence that it was working in 1.0.0, 1,0,1, 1.0.2, 1,1.0, and it there is no point to pretend that it did.

**Comment by t-j-h** [comment]
See #5035 where this discussion continues.

**Comment by romen** [comment]
What about this PR? Should it be merged, closed or put on hold?

**Comment by mspncp** [comment]
It's almost a year now and nobody missed the OS390-Unix target on the 1.1.0 branch. Time to close this ticket?

**Comment by paulidale** [comment]
+1 to close from here.

**Comment by mspncp** [comment]
@t-j-h is it ok for you to close this? The 1.1.0 branch is in security-fix-only mode anyway.

**Comment by paulidale** [comment]
Closing this because it isn't a security fix and it is slated for 1.1.0 only and nobody commented otherwise since December.
